File: services/atlas-tui/internal/tui/readiness_test.go
package tui

import (
	"strings"
	"testing"

	"atlas-tui/internal/client"
)

func TestExecutionReadinessStates(t *testing.T) {
	remediation := "configure an API key or select Claude Code"
	tests := []struct {
		name      string
		status    client.ProviderStatus
		allowMock bool
		kind      readinessKind
		canRun    bool
	}{
		{
			name:   "live",
			status: client.ProviderStatus{Provider: "anthropic", Model: "claude", MockMode: false},
			kind:   readinessLive,
			canRun: true,
		},
		{
			name:   "missing provider projection",
			status: client.ProviderStatus{},
			kind:   readinessUnconfigured,
		},
		{
			name: "unconfigured",
			status: client.ProviderStatus{
				Provider: "openrouter", MockMode: true, Remediation: &remediation,
			},
			kind: readinessUnconfigured,
		},
		{
			name: "degraded",
			status: client.ProviderStatus{
				Provider: "openrouter", MockMode: true, CredentialsPresent: true,
				Remediation: &remediation,
			},
			kind: readinessDegraded,
		},
		{
			name:      "explicit mock",
			status:    client.ProviderStatus{Provider: "mock", MockMode: true},
			allowMock: true,
			kind:      readinessMock,
			canRun:    true,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := readinessFor(tc.status, tc.allowMock)
			if got.Kind != tc.kind || got.CanRun != tc.canRun {
				t.Fatalf("readiness = %+v, want kind=%q canRun=%v", got, tc.kind, tc.canRun)
			}
		})
	}
}

func TestSubmitBlocksImplicitMockAndPreservesDraft(t *testing.T) {
	m := readyModel(120, 36)
	m.status.MockMode = true
	m.status.CredentialsPresent = false
	remediation := "select an available provider"
	m.status.Remediation = &remediation
	m.focus = focusComposer
	m.composer.Focus()
	m.composer.SetValue("analyse the codebase")

	updated, cmd := m.submitComposer()
	got := updated.(model)
	if cmd == nil {
		t.Fatal("blocked submit must load provider settings")
	}
	if got.composer.Value() != "analyse the codebase" {
		t.Fatalf("draft was lost: %q", got.composer.Value())
	}
	if got.focus != focusSettings || !got.settingsLoading {
		t.Fatalf("blocked submit did not open provider settings: focus=%v loading=%v", got.focus, got.settingsLoading)
	}
	if !strings.Contains(got.errMsg, "LIVE PROVIDER REQUIRED") {
		t.Fatalf("missing honest readiness notice: %q", got.errMsg)
	}
}

File: services/atlas-terminal/test/readiness.test.ts
import { describe, expect, it } from 'bun:test';
import { readinessFor, type ProviderStatus } from '../src/tui/util/readiness';

// Mirrors services/atlas-tui/internal/tui/readiness_test.go's cases so both
// TUI surfaces classify provider health identically.
function status(overrides: Partial<ProviderStatus> = {}): ProviderStatus {
	return {
		provider: '',
		model: '',
		auth_mode: '',
		auth_mode_label: '',
		base_url: null,
		credentials_present: false,
		mock_mode: false,
		remediation: null,
		reasoning_effort: null,
		privacy_warning: null,
		...overrides
	};
}

describe('readinessFor', () => {
	it('is LIVE when not mock and provider/model are set', () => {
		const r = readinessFor(status({ provider: 'openrouter', model: 'anthropic/claude-sonnet-4' }), false);
		expect(r).toEqual({ kind: 'live', canRun: true, label: 'LIVE', remediation: '' });
	});

	it('fails closed to UNCONFIGURED on a zero-value projection, never LIVE', () => {
		const r = readinessFor(status(), false);
		expect(r.kind).toBe('unconfigured');
		expect(r.canRun).toBe(false);
	});

	it('is MOCK / EXPLICIT when mock_mode and allowMock are both set', () => {
		const r = readinessFor(status({ mock_mode: true, remediation: 'set an API key' }), true);
		expect(r).toEqual({ kind: 'mock', canRun: true, label: 'MOCK / EXPLICIT', remediation: 'set an API key' });
	});

	it('is DEGRADED when mock but credentials are present and mock is not explicitly allowed', () => {
		const r = readinessFor(status({ mock_mode: true, credentials_present: true }), false);
		expect(r.kind).toBe('degraded');
		expect(r.canRun).toBe(false);
	});

	it('is UNCONFIGURED when mock, no credentials, and mock not allowed', () => {
		const r = readinessFor(status({ mock_mode: true, credentials_present: false }), false);
		expect(r.kind).toBe('unconfigured');
		expect(r.canRun).toBe(false);
	});
});
